### PR TITLE
messaging service for topic function

### PR DESCRIPTION
**Pull request by cdyangzhenyu**
Redmine 10659

Originally, zaqar has no topic function, it is integration with the queue, but our products need to separate the queue and topic, each have different attributes. so I follow the queue to implement the topic function, add some attributes for topic.
This PR mainly to implement the basic api functions of the topic, such as create/list/get/update/delete.

### CODE REVIEW

File: zaqar/storage/base.py [hunk 1]
```diff
@@ -407,6 +417,77 @@ def stats(self, name, project=None):
...
+        :param marker: The last topic name
+        :param limit: (Default 10) Max number of topics to return
+        :param detailed: Whether metadata is included
+
+        :returns: An iterator giving a sequence of topics
+            and the marker of the next page.
+        """
+        return self._list(project, marker, limit, detailed)
+
+    _list = abc.abstractmethod(lambda x: None)
```
Comment (zh-f) [inline zaqar/storage/base.py:444]: 这是干啥的……定义了一个抽象方法，啥也没干？

Comment (cdyangzhenyu) [reply zaqar/storage/base.py:444]: 嗯，这个是个基类，子类里要去实现_list方法。方便不同驱动的实现。

File: zaqar/storage/mongodb/topics.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    field names will be, most of the time, the first
+    letter of their long name.
+"""
+
+from oslo_log import log as logging
+from oslo_utils import timeutils
+import pymongo.errors
+
+from zaqar.common import decorators
+from zaqar.i18n import _
```
Comment (zh-f) [inline zaqar/storage/mongodb/topics.py]: imported but unused. 这个文件跑一下flake8有2出需要空两行，结果空一行；三处缺少空格，在:或者,后。

File: zaqar/storage/pipeline.py [hunk 3]
```diff
@@ -137,6 +137,14 @@ def is_alive(self):
 
     def _health(self):
         return self._storage._health()
+
```
Comment (zh-f) [inline zaqar/storage/pipeline.py]: pipeline.py:140:1: W293 blank line contains whitespace

File: zaqar/storage/pooling.py [hunk 4]
```diff
@@ -270,6 +278,88 @@ def _stats(self, name, project=None):
...
+
+    def _delete(self, name, project=None):
+        control = self._get_controller(name, project)
+        if control:
+            self._pool_catalog.deregister(name, project)
+            ret = control.delete(name, project)
+            return ret
+
+        return None
+
```
Comment (zh-f) [inline zaqar/storage/pooling.py:362]: excepting 2 blank lines, found 1

File: zaqar/transport/validation.py [hunk 5]
```diff
@@ -231,6 +235,19 @@ def _parse_json_schema_change(self, raw_change, draft_version):
 
         path_list = self._decode_json_pointer(path)
         return op, path_list
+
```
Comment (zh-f) [inline zaqar/transport/validation.py]: transport/validation.py:238:1: W293 blank line contains whitespace

File: zaqar/transport/wsgi/v2_0/topics.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+                metadata = wsgi_utils.sanitize(document, spec=None)
+            self._validate.queue_metadata_putting(metadata)
+        except validation.ValidationFailed as ex:
+            LOG.debug(ex)
+            raise wsgi_errors.HTTPBadRequestAPI(six.text_type(ex))
+
+        try:
+            for meta in self._reserved_metadata:
+                if meta not in metadata:
+                    metadata['_%s' % meta] = self._validate.get_limit_conf_value(meta)
```
Comment (zh-f) [inline zaqar/transport/wsgi/v2_0/topics.py]: line too long

File: zaqar/storage/pooling.py (hunk 4 above)
Comment (cdyangzhenyu) [reply zaqar/storage/pooling.py:362]: 恩 跑了一遍flake8 现在应该没问题了

**Review by zh-f — APPROVED** [review]
Fine by me.

File: zaqar/storage/pipeline.py (hunk 3 above)
Comment (blkart) [reply zaqar/storage/pipeline.py]: +1

File: zaqar/transport/wsgi/v2_0/topics.py [hunk 7]
```diff
@@ -40,6 +41,32 @@ def __init__(self, validate, topic_controller, message_controller):
...
+        return reserved_metadata
+
+    @decorators.TransportLog("Topics item")
+    @acl.enforce("topics:get")
+    def on_get(self, req, resp, project_id, topic_name):
+        try:
+            resp_dict = self._topic_controller.get(topic_name,
+                                                   project=project_id)
+        except storage_errors.DoesNotExist as ex:
+            LOG.debug(ex)
```
Comment (blkart) [inline zaqar/transport/wsgi/v2_0/topics.py:59]: 这里确认用 debug ？

File: zaqar/transport/wsgi/v2_0/__init__.py [hunk 8]
```diff
@@ -22,6 +22,7 @@
 from zaqar.transport.wsgi.v2_0 import pools
 from zaqar.transport.wsgi.v2_0 import purge
 from zaqar.transport.wsgi.v2_0 import queues
+from zaqar.transport.wsgi.v2_0 import topics
```
Comment (blkart) [inline zaqar/transport/wsgi/v2_0/__init__.py:25]: 应该放到 subscriptions 下面吧？

**Pull request merged**